src/libopensc/card-cardos.c

```diff
@@ -216,34 +308,99 @@ static int cardos_init(sc_card_t *card)
 		return SC_ERROR_INVALID_CARD;
 	data_field_length = ((rbuf[0] << 8) | rbuf[1]);
 
-	/* strip the length of possible Lc and Le bytes */
-	if (card->caps & SC_CARD_CAP_APDU_EXT)
-		card->max_send_size = data_field_length - 6;
-	else
-		card->max_send_size = data_field_length - 3;
-	/* strip the length of SW bytes */
-	card->max_recv_size = data_field_length - 2;
+	/* TODO is this really needed? strip the length of possible Lc and Le bytes */
+
+	/* Use Min card sizes and reader too. for V5_3 at least*/
+
+	if (card->type == SC_CARD_TYPE_CARDOS_V5_0 || card->type == SC_CARD_TYPE_CARDOS_V5_3) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "data_field_length:%"SC_FORMAT_LEN_SIZE_T"u "
+				"card->reader->max_send_size:%"SC_FORMAT_LEN_SIZE_T"u "
+				"card->reader->max_recv_size:%"SC_FORMAT_LEN_SIZE_T"u %s",
+				data_field_length, card->reader->max_send_size, card->reader->max_recv_size,
+				(card->caps & SC_CARD_CAP_APDU_EXT) ? "SC_CARD_CAP_APDU_EXT" : " ");
```

Comment:
clang-format suggested the following change:


```
-		sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "data_field_length:%"SC_FORMAT_LEN_SIZE_T"u "
-				"card->reader->max_send_size:%"SC_FORMAT_LEN_SIZE_T"u "
-				"card->reader->max_recv_size:%"SC_FORMAT_LEN_SIZE_T"u %s",
-				data_field_length, card->reader->max_send_size, card->reader->max_recv_size,
-				(card->caps & SC_CARD_CAP_APDU_EXT) ? "SC_CARD_CAP_APDU_EXT" : " ");
+		sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "data_field_length:%" SC_FORMAT_LEN_SIZE_T "u "
+							 "card->reader->max_send_size:%" SC_FORMAT_LEN_SIZE_T "u "
+							 "card->reader->max_recv_size:%" SC_FORMAT_LEN_SIZE_T "u %s",
+			 data_field_length, card->reader->max_send_size, card->reader->max_recv_size,
+			 (card->caps & SC_CARD_CAP_APDU_EXT) ? "SC_CARD_CAP_APDU_EXT" : " ");
```

<!-- identifier: clang-format -->
<!-- File: src/libopensc/card-cardos.c -->